tactics/hints.ml

```diff
@@ -154,12 +154,18 @@ type hint_pattern =
 | DefaultPattern
 | ConstrPattern of constr_pattern
 
+type matching_mode =
+| SyntacticMatching
+| TransparencyMatching
+
 type 'a with_metadata =
-  { pri     : int
+  { pri      : int
   (** A number lower is higher priority *)
-  ; pat     : hint_pattern option
+  ; pat      : hint_pattern option
   (** A pattern for the concl of the Goal *)
-  ; name    : GlobRef.t option
+  ; matching : matching_mode
```

Comment:
The matching field should be merged with the pat field, by duplicating the ConstrPattern constructor from the hint_pattern type, e.g. as `SyntacticPattern of constr_pattern`. This would rule out statically some impossible cases.

Reply:
I went with something slightly different because I misread your comment. Let me know if this is OK or if you want separate constructors instead.

Reply:
I pushed a new version with a separate constructor. I couldn't find a very descriptive name for the existing one that would explain what it does so I kept `ConstrPattern`.